src/coreclr/jit/importer.cpp

```diff
@@ -4503,13 +4503,135 @@ GenTree* Compiler::impIntrinsic(GenTree*                newobjThis,
+
+                    if (ni == NI_System_Math_Max)
+                    {
+                        cnsNode->gtDconVal =
+                            FloatingPointUtils::maximum(cnsNode->gtDconVal, otherNode->AsDblCon()->gtDconVal);
+                    }
+                    else
+                    {
+                        assert(ni == NI_System_Math_Min);
+                        cnsNode->gtDconVal =
+                            FloatingPointUtils::minimum(cnsNode->gtDconVal, otherNode->AsDblCon()->gtDconVal);
+                    }
+
+                    retNode = cnsNode;
+                    break;
+                }
+
+                // only one is constant, we can fold in specialized scenarios
+
+                if (cnsNode->IsFloatNaN())
+                {
+                    // maxsd, maxss, minsd, and minss all return op2 if either is NaN
+                    // we require NaN to be propagated so ensure the known NaN is op2
+
+                    op2 = cnsNode;
+                    op1 = otherNode;
+                }
+                else if (ni == NI_System_Math_Max)
+                {
+                    if (cnsNode->IsFloatPositiveZero())
+                    {
+                        // maxsd, maxss return op2 if both inputs are 0 of either sign
+                        // we require +0 to be greater than -0, so we can't handle if
+                        // the known constant is +0. This is because if the unknown value
+                        // is -0, we'd need the cns to be op2. But if the unknown value
+                        // is NaN, we'd need the cns to be op1 instead.
+
+                        break;
+                    }
+
+                    // Given the checks, op1 can safely be the cns and op2 the other node
+                    ni = (callType == TYP_DOUBLE) ? NI_SSE2_Max : NI_SSE_Max;
+
+                    op1 = cnsNode;
+                    op2 = otherNode;
+                }
+                else
+                {
+                    assert(ni == NI_System_Math_Min);
+
+                    if (cnsNode->IsFloatNegativeZero())
+                    {
+                        // minsd, minss return op2 if both inputs are 0 of either sign
+                        // we require -0 to be lesser than +0, so we can't handle if
+                        // the known constant is -0. This is because if the unknown value
+                        // is +0, we'd need the cns to be op2. But if the unknown value
+                        // is NaN, we'd need the cns to be op1 instead.
+
+                        break;
+                    }
+
+                    // Given the checks, op1 can safely be the cns and op2 the other node
+                    ni = (callType == TYP_DOUBLE) ? NI_SSE2_Min : NI_SSE_Min;
+
+                    op1 = cnsNode;
+                    op2 = otherNode;
+                }
+
+                // one is constant and we know its something we can handle, so pop both peeked values
```

Comment:
We should probably assert that `op1->IsCnsFltOrDbl() && !op2->IsCnsFltOrDbl()`

Reply:
I'm not sure about the exact line of the assert to put on. If we put it around 4620 then it doesn't seem right, because op2 can be const as well, can't it?
```C#
if (cnsNode->IsFloatNaN())
{
   // maxsd, maxss, minsd, and minss all return op2 if either is NaN
   // we require NaN to be propagated so ensure the known NaN is op2

    op2 = cnsNode;
    op1 = otherNode;
}
```

Reply:
Awesome, that should work 👍

Reply:
If the non-constant node has side effects that will discard them. I guess the easiest though somewhat conservative thing is to call `impSpillSideEffects` with `CHECK_SPILL_ALL` before that.

Reply:
It is a good idea, I'm trying to figure out how `impSpillSideEffects` works by reading the summary of the function. It looks simple and tricky at the same time.
```C#
/*****************************************************************************
 *
 *  If the stack contains any trees with side effects in them, assign those
 *  trees to temps and replace them on the stack with refs to their temps.
 *  [0..chkLevel) is the portion of the stack which will be checked and spilled.
 */
```
So I use this call prior `impPopStack();` and `DEBUG_DESTROY_NODE(otherNode);` and it helps to presave the side effects because the refs will be wiped instead of original trees (in case there were any)?

Reply:
Correct. The importer maintains both the stack of nodes (which may have side effects) and a list of statements in order of execution. `impSpillSideEffects` will evaluate side effecting nodes into locals in order and add the created assignments to the list of statements, while replacing the stack entries with references to those new locals.
You can actually just use `impSpillSideEffect` (singular) to spill the single entry (and previous interfering ones) here.